Jan 2020 milestone 2️⃣0️⃣2️⃣0️⃣

dhmlau opened the pull request:
# Jan milestone

## Overview

- Take some time to:
  - fix bugs, improve developer experience, fix docs
  - triage open issues across all the repos, not just `loopback-next`
- Migration work continues to be one of the focus
- Focus on issues/PRs for loopback-connector-mongodb

## Goals

- Migration guide between LB3 and LB4 MVP #453

  - [ ] How to migrate boot scripts #3957

- Improve developer experience

  - [ ] Expose isActive() method in Transaction interface #3471
  - [ ] Missing model.ts after running lb4 openapi command #4135
  - [ ] Model strict: false is ignored with Postgre SQL repository #4042

- Bugs

  - [ ] :running: Docs: Improve TodoList Tutorial instructions and completed tutorial
        'lb4 example todo-list' #4161
  - [x] lb3 - Relations on uuid
        https://github.com/strongloop/loopback/issues/4287

- Create models & REST APIs dynamically at runtime

  - [ ] [Needs estimates] Dynamic binding/rebinding of controllers after app
        start #433

- Update dependencies

  - [ ] [Needs estimates] chore: update dependency change-case to v4 #4233
  - [ ] [Needs estimates] update dependency ts-morph to v6 #4274

- [ ] [spike] API Connect / LoopBack 4 integration #4115

- _Miscellaneous_
  - [ ] Spike: Migration guide from LB3 - Authentication & authorization
#3719
  - [ ] :running: API Explorer ignores filter options where, fields, and order #2208
  - [ ] lb4 repository,model,controller commands generate artifacts with 'lint'
issues #4205

## Stretch goals

- Migration guide between LB3 and LB4 MVP #453

  - [ ] How to migrate user-defined model methods #3949
  - [ ]  How to migrate remoting hooks #3950

- Create models & REST APIs dynamically at runtime

  - [ ] How to build models, repositories and controllers dynamically at runtime
        #4296 [needs estimate]

- Bugs

  - [ ] [community contribution?] Preserve custom type of auto-generated id
    property #3602 [needs estimate]
  - [ ] Column names in lowercase using lb4 discover method -
        SQL connector #3343
  - [ ] :running: lb4 relation should discard changes when something failed #3287 

- PR reviews
  - [ ] feat(extension-logging): add integration with winston and fluentd
        logging, https://github.com/strongloop/loopback-next/pull/4117
  - [ ] [RFC] feat: add base code for the image of Appsody LoopBack stack,
        https://github.com/strongloop/loopback-next/pull/4250 [draft pr still in discussion]

bajtos commented on docs/_jan-milestone_.md:
```diff
@@ -0,0 +1,50 @@
+
+  - [ ] [3] Docs: Improve TodoList Tutorial instructions and completed tutorial
+        'lb4 example todo-list' #4161
+
```
I am proposing to add the following bug(s) to the Jan milestone:

- https://github.com/strongloop/loopback/issues/4287 lb3 - Relations on uuid

bajtos replied:
- Preserve custom type of auto-generated id property #3602

dhmlau replied:
for #3602, will put it in the stretch goal. It seems like @frbuceta will pick it up.

bajtos commented on docs/_jan-milestone_.md:
```diff
@@ -0,0 +1,50 @@
+  - [ ] [3]Example app showing CrudRestApiBuilder #3738
+
+- [ ] [3][spike] API Connect / LoopBack 4 integration #4115
+
```
This dependency upgrade requires deeper investigation, let's plan some time in January to get it done.

- chore: update dependency change-case to v4 #4233

bajtos replied:
Another similar one:

- chore: update dependency ts-morph to v6 #4274

dhmlau commented:
@bajtos, I've incorporated your comments and also moved some of the migration guide tasks to stretch goal. We currently are committing 26 points + 2 tasks need to be estimated.  It would be a good committed number of story points due to vacation time and we can have more flexibility to fix bugs/usability issues.

dhmlau commented:
@bajtos, I've updated the Jan milestone based on our discussion today:
- have more focus / people to work on the migration story 
- remove the "from model definition to REST" epic / tasks. 

Could you please add the migration tasks that you think we should work on in Jan? Thanks.

dhmlau commented on docs/_jan-milestone_.md:
```diff
@@ -0,0 +1,57 @@
+
+- [ ] [3][spike] API Connect / LoopBack 4 integration #4115
+
+- [ ] [Needs estimates] chore: update dependency change-case to v4 #4233
```
similarly, we can add chore: update dependency ts-morph to v6 https://github.com/strongloop/loopback-next/pull/4274?

bajtos replied:
Yes please!

bajtos commented on docs/_jan-milestone_.md:
```diff
@@ -0,0 +1,57 @@
+
+## Stretch goals
+
+- [ ] Spike: migrate tests to Jest (ts-jest) #3159
```
Let's remove this item. Based on my discussion with @boneskull, we would like to try few experiments with Mocha to see if we can get DX that's good enough.

bajtos commented on docs/_jan-milestone_.md:
```diff
@@ -0,0 +1,57 @@
+  - [ ] [3] Expose isActive() method in Transaction interface #3471
+  - [ ] [3] Missing model.ts after running lb4 openapi command #4135
+  - [ ] [2] Model strict: false is ignored with Postgre SQL repository #4042
+
```
IIRC, we wanted to also add one or two tasks for defining models at runtime, see https://github.com/strongloop/loopback-next/pull/4116/files#r357149740

```markdown
#### Create models & REST APIs dynamically at runtime

- Dynamic binding/rebinding of controllers after app start #433
- How to build models, repositories and controllers dynamically at runtime #4296
```

I am proposing to commit the first task for January and the second for February. Thoughts?

dhmlau replied:
I've put #433 to committed stories in Jan, and #4296 as stretch goal.

dhmlau commented:
https://github.com/strongloop/loopback-next/issues/4376 is opened. closing this PR. Thanks!
